rfd/0167-console-service.md

```diff
@@ -0,0 +1,194 @@
+
+	server := http.Server{Handler: mux}
+	server.Serve(listener)
+}
+
+// The client however, requires a different transport. The differences is on the
+// DialContext implementation.
+func createClient(socketAddr string) *http.Client {
+	return &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("unix", socketAddr)
+			},
+		}
+	}
+}
+
+// While making requests using the client, the users will still require to
+// provide the full URL since the client always validate them.
+res, err := client.Get("http://console-service/debug/pprof/heap")
+```
+</details>
+
+#### Changing log level command
+
+Changes the application log level.
+
+`teleport [-c config-path] console set-log-level [LEVEL]`
+
+|Flag|Description|Default value|
+|----|-----------|-------------|
+|`-c,--config`|Teleport configuration path.|`/etc/teleport.yaml`|
+|`LEVEL`|Log level (case-insensitive). Any of: `DEBUG`, `INFO`, `WARN`, `ERROR`|``|
+
+Usage examples:
+```bash
+$ teleport console set-log-level DEBUG
+$ teleprot -c /random/teleport.yaml console set-log-level INFO
+```
+
+#### Capture `pprof` profiles command
+
+Export the application profile (`pprof` format).
+
+`teleport [-c config-path] console profile [--seconds=] [PROFILE_NAME]`
+
+The `PROFILE_NAME` values follow the Golang's definition on `runtime.Profile`
+plus the profiles defined by `net/http/pprof`:
+- `allocs`: A sampling of all past memory allocations
+- `block`: Stack traces that led to blocking on synchronization primitives
+- `cmdline`: The command line invocation of the current program
+- `goroutine`: Stack traces of all current goroutines. Use debug=2 as a query parameter to export in the same format as an unrecovered panic.
+- `heap`: A sampling of memory allocations of live objects. You can specify the gc GET parameter to run GC before taking the heap sample.
+- `mutex`: Stack traces of holders of contended mutexes
+- `profile`: CPU profile. You can specify the duration in the seconds GET parameter. After you get the profile file, use the go tool pprof command to investigate the profile.
+- `threadcreate`: Stack traces that led to the creation of new OS threads
+- `trace`: A trace of execution of the current program. You can specify the duration in the seconds GET parameter. After you get the trace file, use the go tool trace command to investigate the trace.
+
+Note: `--seconds` argument has the same effect as [`seconds` query string](https://pkg.go.dev/net/http/pprof#hdr-Parameters).
+
+|Flag|Description|Default value|
+|----|-----------|-------------|
+|`-c,--config`|Teleport configuration path.|`/etc/teleport.yaml`|
+|`-s,--seconds`|For CPU and trace profiles, profile for the given duration. For other profiles, return a delta profile.|None|
+|`PROFILE_NAME`|Profile to be exported. Any of: `allocs`, `block`, `cmdline`, `goroutine`, `heap`, `mutex`, `profile`, `threadcreate`, `trace`|Required|
+
+### Security
```

Comment:
We probably don't want to take advantage of this - but I thought I'd mention that it is possible with UDS to determine the UID/GID and PID of the calling process. This can give a little bit more control/insight than just using FS permissions to mange access to a socket.

Reply:
That's interesting. I would require an additional configuration, but we could specify which UID/GID can perform those commands. I'll add a mention to it on the RFD.